File: computing/interview/design-patterns/behavioral/chain-of-responsibility/csharp/ChainOfResponsibilityDemo.cs
using System;

namespace Interview.DesignPatterns.Behavioral.ChainOfResponsibility
{
    public enum Severity
    {
        Info,
        Warning,
        Error
    }

    public readonly struct LogRecord
    {
        public LogRecord(Severity level, string message)
        {
            Level = level;
            Message = message;
        }

        public Severity Level { get; }
        public string Message { get; }
    }

    public abstract class Logger
    {
        private Logger? _next;

        public void SetNext(Logger next)
        {
            _next = next;
        }

        public void Handle(LogRecord record)
        {
            if (ShouldHandle(record.Level))
            {
                Write(record.Message);
            }
            else if (_next != null)
            {
                _next.Handle(record);
            }
            else
            {
                Console.WriteLine($"No handler for message: {record.Message}");
            }
        }

        protected abstract bool ShouldHandle(Severity level);
        protected abstract void Write(string message);
    }

    public sealed class ConsoleLogger : Logger
    {
        protected override bool ShouldHandle(Severity level) => level == Severity.Info;
        protected override void Write(string message) => Console.WriteLine($"Console: {message}");
    }

    public sealed class FileLogger : Logger
    {
        protected override bool ShouldHandle(Severity level) => level == Severity.Warning;
        protected override void Write(string message) => Console.WriteLine($"File: {message}");
    }

    public sealed class AlertLogger : Logger
    {
        protected override bool ShouldHandle(Severity level) => level == Severity.Error;
        protected override void Write(string message) => Console.WriteLine($"Alert: {message}");
    }

    public static class Demo
    {
        public static void Main()
        {
            var console = new ConsoleLogger();
            var file = new FileLogger();
            var alert = new AlertLogger();

            console.SetNext(file);
            file.SetNext(alert);

            console.Handle(new LogRecord(Severity.Info, "Starting system"));
            console.Handle(new LogRecord(Severity.Warning, "Disk space low"));
            console.Handle(new LogRecord(Severity.Error, "Service offline"));
        }
    }
}

File: computing/interview/design-patterns/behavioral/chain-of-responsibility/java/ChainOfResponsibilityDemo.java
package interview.designpatterns.behavioral.chainofresponsibility;

enum Severity {
    INFO,
    WARNING,
    ERROR
}

final class LogRecord {
    private final Severity level;
    private final String message;

    LogRecord(Severity level, String message) {
        this.level = level;
        this.message = message;
    }

    Severity level() {
        return level;
    }

    String message() {
        return message;
    }
}

abstract class Logger {
    private Logger next;

    void setNext(Logger next) {
        this.next = next;
    }

    void handle(LogRecord record) {
        if (shouldHandle(record.level())) {
            write(record.message());
        } else if (next != null) {
            next.handle(record);
        } else {
            System.out.printf("No handler for message: %s%n", record.message());
        }
    }

    protected abstract boolean shouldHandle(Severity level);
    protected abstract void write(String message);
}

final class ConsoleLogger extends Logger {
    @Override
    protected boolean shouldHandle(Severity level) {
        return level == Severity.INFO;
    }

    @Override
    protected void write(String message) {
        System.out.printf("Console: %s%n", message);
    }
}

final class FileLogger extends Logger {
    @Override
    protected boolean shouldHandle(Severity level) {
        return level == Severity.WARNING;
    }

    @Override
    protected void write(String message) {
        System.out.printf("File: %s%n", message);
    }
}

final class AlertLogger extends Logger {
    @Override
    protected boolean shouldHandle(Severity level) {
        return level == Severity.ERROR;
    }

    @Override
    protected void write(String message) {
        System.out.printf("Alert: %s%n", message);
    }
}

public final class ChainOfResponsibilityDemo {
    private ChainOfResponsibilityDemo() {}

    public static void main(String[] args) {
        Logger console = new ConsoleLogger();
        Logger file = new FileLogger();
        Logger alert = new AlertLogger();

        console.setNext(file);
        file.setNext(alert);

        console.handle(new LogRecord(Severity.INFO, "Starting system"));
        console.handle(new LogRecord(Severity.WARNING, "Disk space low"));
        console.handle(new LogRecord(Severity.ERROR, "Service offline"));
    }
}
